### PR TITLE
Add the endpoint used by the handshake process to connector's pollset_set

### PR DESCRIPTION
**Pull request by y-zeng**
Before starting the handshake process, there's no guarantee that the endpoint has already been added to the connector's pollset_set (e.g. the endpoint is created from an unix domain socket).  In this case, no completion queue will be driving the handshake process.

### CODE REVIEW

File: src/core/ext/transport/chttp2/client/chttp2_connector.c [hunk 1]
```diff
@@ -136,6 +136,8 @@ static void start_handshake_locked(grpc_exec_ctx *exec_ctx,
   c->handshake_mgr = grpc_handshake_manager_create();
   grpc_handshakers_add(exec_ctx, HANDSHAKER_CLIENT, c->args.channel_args,
                        c->handshake_mgr);
+  grpc_endpoint_add_to_pollset_set(exec_ctx, c->endpoint,
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/client/chttp2_connector.c]: Do we need to also remove this endpoint from interested_parties when handshaking is complete?

**Comment by markdroth** [comment]
Are those bm_diff numbers spurious?  I wouldn't think that a handshaker change would affect the fast path at all, since it only gets run during connection establishment.

**Comment by ncteisen** [comment]
I'd like to see bm_diff run again. If those numbers are real, then we'd have an issue. I see it only happens with unary (streaming actually gets faster in three cases). Would this cause more fds to exist in the polling island for the rest of the program's life? this could be a problem that occurs with epollsig and not epoll1

**Comment by y-zeng** [comment]
For AF_INET and AF_INET6 addresses, the endpoint fd has already been added to the same pollset_set in https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/tcp_client_posix.c#L305. This should only affect unix domain socket addresses. I'll run bm_diff again to check if the numbers will be the same.

File: src/core/ext/transport/chttp2/client/chttp2_connector.c (hunk 1 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/client/chttp2_connector.c]: I think we should remove endpoints from the pollset_set after the handshaking process is complete. But we did't do this in other processes (e.g. establishing tcp connection, creating connected subchannels). And we don't even have an interface to do so. I'll check if there is a reason not doing this. If not I'll add a `grpc_endpoint_del_from_pollset_set` interface.

**Comment by y-zeng** [comment]
Jenkins, test this please.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.0%    +272 [None]                                                   +1.88Ki  +0.0%
      +0.1%    +256 [Unmapped]                                               +1.86Ki  +0.0%
      +1.6%     +16 vtable                                                       +16  +1.6%
  +1.9%     +32 src/core/ext/transport/chttp2/client/chttp2_connector.cc     +32  +1.9%
      +3.2%     +16 connected                                                    +16  +3.2%
      +3.3%     +15 on_handshake_done                                            +15  +3.3%
      +2.0%      +1 [Unmapped]                                                    +1  +2.0%
   +12%     +16 src/core/lib/iomgr/endpoint.cc                               +16   +12%
       +12%     +10 [Unmapped]                                                   +10   +12%
      [NEW]      +6 grpc_endpoint_delete_from_pollset_set                         +6  [NEW]
  +0.2%     +16 src/core/lib/iomgr/tcp_posix.cc                              +16  +0.2%
      [NEW]     +15 tcp_delete_from_pollset_set                                  +15  [NEW]
      +0.6%      +1 [Unmapped]                                                    +1  +0.6%
  +0.4%     +16 src/core/lib/security/transport/secure_endpoint.cc           +16  +0.4%
      [NEW]      +9 endpoint_delete_from_pollset_set                              +9  [NEW]
      +7.9%      +7 [Unmapped]                                                    +7  +7.9%

  +0.0%    +352 TOTAL                                                    +1.95Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.0%    +272 [None]                                                   +1.88Ki  +0.0%
      +0.1%    +256 [Unmapped]                                               +1.86Ki  +0.0%
      +1.6%     +16 vtable                                                       +16  +1.6%
  +1.9%     +32 src/core/ext/transport/chttp2/client/chttp2_connector.cc     +32  +1.9%
      +3.2%     +16 connected                                                    +16  +3.2%
      +3.3%     +15 on_handshake_done                                            +15  +3.3%
      +2.0%      +1 [Unmapped]                                                    +1  +2.0%
   +12%     +16 src/core/lib/iomgr/endpoint.cc                               +16   +12%
       +12%     +10 [Unmapped]                                                   +10   +12%
      [NEW]      +6 grpc_endpoint_delete_from_pollset_set                         +6  [NEW]
  +0.2%     +16 src/core/lib/iomgr/tcp_posix.cc                              +16  +0.2%
      [NEW]     +15 tcp_delete_from_pollset_set                                  +15  [NEW]
      +0.6%      +1 [Unmapped]                                                    +1  +0.6%
  +0.4%     +16 src/core/lib/security/transport/secure_endpoint.cc           +16  +0.4%
      [NEW]      +9 endpoint_delete_from_pollset_set                              +9  [NEW]
      +7.9%      +7 [Unmapped]                                                    +7  +7.9%

  +0.0%    +352 TOTAL                                                    +1.95Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.0%    +272 [None]                                                   +1.88Ki  +0.0%
      +0.1%    +256 [Unmapped]                                               +1.86Ki  +0.0%
      +1.6%     +16 vtable                                                       +16  +1.6%
  +1.9%     +32 src/core/ext/transport/chttp2/client/chttp2_connector.cc     +32  +1.9%
      +3.2%     +16 connected                                                    +16  +3.2%
      +3.3%     +15 on_handshake_done                                            +15  +3.3%
      +2.0%      +1 [Unmapped]                                                    +1  +2.0%
   +12%     +16 src/core/lib/iomgr/endpoint.cc                               +16   +12%
       +12%     +10 [Unmapped]                                                   +10   +12%
      [NEW]      +6 grpc_endpoint_delete_from_pollset_set                         +6  [NEW]
  +0.2%     +16 src/core/lib/iomgr/tcp_posix.cc                              +16  +0.2%
      [NEW]     +15 tcp_delete_from_pollset_set                                  +15  [NEW]
      +0.6%      +1 [Unmapped]                                                    +1  +0.6%
  +0.4%     +16 src/core/lib/security/transport/secure_endpoint.cc           +16  +0.4%
      [NEW]      +9 endpoint_delete_from_pollset_set                              +9  [NEW]
      +7.9%      +7 [Unmapped]                                                    +7  +7.9%

  +0.0%    +352 TOTAL                                                    +1.95Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.0%    +240 [None]                                                   +2.05Ki  +0.0%
      +0.0%    +224 [Unmapped]                                               +2.03Ki  +0.0%
      +1.6%     +16 vtable                                                       +16  +1.6%
  +1.9%     +32 src/core/ext/transport/chttp2/client/chttp2_connector.cc     +32  +1.9%
      +3.2%     +16 connected                                                    +16  +3.2%
      +3.3%     +15 on_handshake_done                                            +15  +3.3%
      +2.0%      +1 [Unmapped]                                                    +1  +2.0%
   +12%     +16 src/core/lib/iomgr/endpoint.cc                               +16   +12%
       +12%     +10 [Unmapped]                                                   +10   +12%
      [NEW]      +6 grpc_endpoint_delete_from_pollset_set                         +6  [NEW]
  +0.2%     +16 src/core/lib/iomgr/tcp_posix.cc                              +16  +0.2%
      [NEW]     +15 tcp_delete_from_pollset_set                                  +15  [NEW]
      +0.6%      +1 [Unmapped]                                                    +1  +0.6%
  +0.4%     +16 src/core/lib/security/transport/secure_endpoint.cc           +16  +0.4%
      [NEW]      +9 endpoint_delete_from_pollset_set                              +9  [NEW]
      +7.9%      +7 [Unmapped]                                                    +7  +7.9%

  +0.0%    +320 TOTAL                                                    +2.12Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++

 -------------- SHRINK --------------
  [ = ]       0 [None]    -120  -0.0%

  [ = ]       0 TOTAL     -120  -0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.0%    +240 [None]                                                   +2.05Ki  +0.0%
      +0.0%    +224 [Unmapped]                                               +2.03Ki  +0.0%
      +1.6%     +16 vtable                                                       +16  +1.6%
  +1.9%     +32 src/core/ext/transport/chttp2/client/chttp2_connector.cc     +32  +1.9%
      +3.2%     +16 connected                                                    +16  +3.2%
      +3.3%     +15 on_handshake_done                                            +15  +3.3%
      +2.0%      +1 [Unmapped]                                                    +1  +2.0%
   +12%     +16 src/core/lib/iomgr/endpoint.cc                               +16   +12%
       +12%     +10 [Unmapped]                                                   +10   +12%
      [NEW]      +6 grpc_endpoint_delete_from_pollset_set                         +6  [NEW]
  +0.2%     +16 src/core/lib/iomgr/tcp_posix.cc                              +16  +0.2%
      [NEW]     +15 tcp_delete_from_pollset_set                                  +15  [NEW]
      +0.6%      +1 [Unmapped]                                                    +1  +0.6%
  +0.4%     +16 src/core/lib/security/transport/secure_endpoint.cc           +16  +0.4%
      [NEW]      +9 endpoint_delete_from_pollset_set                              +9  [NEW]
      +7.9%      +7 [Unmapped]                                                    +7  +7.9%

  +0.0%    +320 TOTAL                                                    +2.12Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++

 -------------- SHRINK --------------
  [ = ]       0 [None]    -120  -0.0%

  [ = ]       0 TOTAL     -120  -0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by y-zeng** [comment]
@markdroth The backup poller (#12732) is blocked by this PR. Could you please take a look?
The secure handshake process was driven by the connectivity watcher. After increasing the poll interval in #12732, we started seeing handshake failures.

**Comment by y-zeng** [comment]
Thanks for the review!
Known issues:
Asan C: #13124 
Asan C++: #13081, #13080
Basic Tests C++ Linux [dbg]: #13081
Basic Tests C++ Linux [opt]: #13081, #13080
Basic Tests MacOS [dbg]: #13131, #13085, #13134
Basic Tests MacOS [opt]: #13131, #13085, #13134
Basic Tests Multi-language Linux: #12886, #12740
Basic Tests Windows [dbg]: #13091, #13079
Basic Tests Windows [opt]: #13091, #13079
Tsan C: #13124 
Tsan C++: #13080

**Pull request merged**